pkg/kubelet/kubeletconfig/util/files/files_test.go

```diff
@@ -0,0 +1,383 @@
+					}
+				}
+			}
+		})
+
+	}
+}
+
+func TestEnsureDir(t *testing.T) {
+	prePath := getTempDir(t, "TestEnsureDir")
+	fs := utilfs.DefaultFs{}
+	defer cleanup(t, fs, prePath)
+
+	testcases := map[string]struct {
+		path        string // path to be created
+		isFile      bool   // if the path indicates a file
+		create      bool   // if should create dir before test
+		expectedErr error
+	}{
+		"exists as file": {
+			path:        prePath + "/ExistsAsFile",
+			isFile:      true,
+			create:      true,
+			expectedErr: fmt.Errorf(""), // should have error, but the error info can only be fulfilled after creating the file
+		},
+		"exists as dir": {
+			path:        prePath + "/NotExists",
+			isFile:      false,
+			create:      true,
+			expectedErr: nil,
+		},
+		//"not exists in sub dir": {
+		//	path:        prePath + "/SubDir1/SubDir2/Exists",
+		//	isFile:      false,
+		//	create:      false,
+		//	expectedErr: nil,
+		//},
+	}
+
+	for testcaseName, testcase := range testcases {
+		var err error
+		t.Run(testcaseName, func(t *testing.T) {
+			if testcase.create {
+				if testcase.isFile {
+					err = createFile(fs, testcase.path)
+				} else {
+					err = createDir(fs, testcase.path)
+				}
+			}
+
+			// not nil indicates we should specify the error
+			if testcase.expectedErr != nil {
+				testcase.expectedErr = fmt.Errorf("expected dir at %q, but mode is %q", testcase.path, getFileMode(fs, testcase.path))
+			}
+
+			if err != nil {
+				t.Errorf("unexpected error in creating dir: %s, error: %v", testcase.path, err)
+			} else {
+				// ensure dir exists
+				if err = EnsureDir(fs, testcase.path); err != nil {
+					// if the path is file, should get error
+					if !reflect.DeepEqual(testcase.expectedErr, err) {
+						t.Errorf("unexpected error in creating dir, expected: %v, actual: %v", testcase.expectedErr, err)
+					}
+				} else {
+					// check if dir created
+					info, err := fs.Stat(testcase.path)
+					if err != nil {
+						t.Errorf("unexpected error in verifying dir: %s, error: %v", testcase.path, err)
+					}
+					if !info.IsDir() {
+						t.Errorf("expected dir in path: %q, but got: %v", testcase.path, info.Mode())
+					}
+				}
+			}
+		})
+	}
+}
+
+func getTempDir(t *testing.T, subPath string) string {
+	dir, err := ioutil.TempDir("", subPath)
+	if err != nil {
+		t.Fatalf("error in getting temp dir via ioutil, error: %v", err)
+	}
+	return dir
+}
+
+func getFileMode(fs utilfs.Filesystem, path string) string {
+	if info, err := fs.Stat(path); err == nil {
+		return info.Mode().String()
+	}
+	return ""
+}
+
+func createFile(fs utilfs.Filesystem, path string) error {
+	if err := createDir(fs, filepath.Dir(path)); err != nil {
+		return err
+	}
+
+	// create the file
+	if _, err := fs.Create(path); err != nil && !os.IsExist(err) {
+		return err
+	}
+
+	return nil
+}
+
+func createDir(fs utilfs.Filesystem, path string) error {
+	if err := fs.MkdirAll(path, defaultPerm); err != nil && !os.IsExist(err) {
```

Comment:
I think this line is the issue - you're creating the `FileNotExistsInSubDir` directory with permission `0666` (ref:
https://github.com/kubernetes/kubernetes/blob/4b41a54eda7375cc13175acbd8d58146d207d123/pkg/kubelet/kubeletconfig/util/files/files.go#L27)

Since this is lacking the execute bit, you're not able to create a file inside it later.

Reply:
I'm pretty sure this bug exists in `EnsureDir` and `EnsureFile` too (from `files.go`), which is exciting.

Reply:
I guess your failure is actually on `EnsureDir`, so yay test cases.

Reply:
@ixdy thanks for you review! so what do suggest for the permission, is 0755 ok? Also, do you mean that the 0666 permission would prevent the invokers to create files or dirs under the underlying dir? If so, how about changing the defaultPerm permission to 0755?

Reply:
Maybe we should actually use 0700 for directories and 0666 for files, since these are the defaults in the go std lib? @ixdy wdyt?